File: javascript/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const bracket = {
    "(": ")",
    "[": "]",
    "{": "}",
  };
  const stack = [];
  for (let c of s) {
    if (c === "(" || c === "[" || c === "{") stack.push(c);
    else if (bracket[stack.pop()] != c) return false;
  }
  return stack.length === 0;
};
// @lc code=end

File: rust/20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.len() % 2 != 0 {
            return false;
        }

        let mut stack: Vec<char> = Vec::with_capacity(s.len() / 2);
        for c in s.chars() {
            match c {
                '(' | '[' | '{' => stack.push(c),
                _ => match stack.pop() {
                    Some('(') if c == ')' => (),
                    Some('[') if c == ']' => (),
                    Some('{') if c == '}' => (),
                    _ => return false,
                },
            }
        }
        return stack.is_empty();
    }
}
// @lc code=end
